### PR TITLE
ipam/crd: Fix spurious CiliumNode update status failures

### PR DESCRIPTION
**Pull request by gandro**
When running in CRD-based IPAM modes (Alibaba, Azure, ENI, CRD), it is
possible to observe spurious "Unable to update CiliumNode custom
resource" failures in the cilium-agent.

The full error message is as follows: "Operation cannot be fulfilled on
ciliumnodes.cilium.io <node>: the object has been modified; please apply
your changes to the latest version and try again".

It means that the Kubernetes `UpdateStatus` call has failed because the
local `ObjectMeta.ResourceVersion` of submitted CiliumNode version is
out of date. In the presence of races, this error is expected and will
resolve itself once the agent receives a more recent version of the
object with the new resource version.

However, it is possible that the resource version of a `CiliumNode`
object is bumped even though the `Spec` or `Status` of the `CiliumNode`
remains the same. This for examples happens when
`ObjectMeta.ManagedFields` is updated by the Kubernetes apiserver.

Unfortunately, `CiliumNode.DeepEqual` does _not_ consider any
`ObjectMeta` fields (including the resource version). Therefore two
objects with different resource versions are considered the same by the
`CiliumNode` watcher used by IPAM.

But to be able to successfully call `UpdateStatus` we need to know the
most recent resource version. Otherwise, `UpdateStatus` will always fail
until the `CiliumNode` object is updated externally for some reason.

Therefore, this commit modifies the logic to always store the most
recent version of the `CiliumNode` object, even if `Spec` or `Status`
has not changed.  This in turn allows `nodeStore.refreshNode` (which
invokes `UpdateStatus`) to always work on the most recently observed
resource version.

```release-note
ipam/crd: Fix spurious "Unable to update CiliumNode custom resource" failures in cilium-agent
```

### CODE REVIEW

File: pkg/ipam/crd.go [hunk 1]
```diff
@@ -128,12 +128,8 @@ func newNodeStore(nodeName string, conf Configuration, owner Owner, k8sEventReg
 				defer func() { k8sEventReg.K8sEventReceived("CiliumNode", "update", valid, equal) }()
 				if oldNode, ok := oldObj.(*ciliumv2.CiliumNode); ok {
 					if newNode, ok := newObj.(*ciliumv2.CiliumNode); ok {
-						if oldNode.DeepEqual(newNode) {
-							equal = true
-							return
-						}
 						valid = true
```
Comment (twpayne) [inline pkg/ipam/crd.go:131]: It looks like there was another bug in the old code here: `valid` was only set to `true` if `oldNode` and `newNode` were not equal. This PR looks like it fixes that.

File: pkg/ipam/crd.go [hunk 2]
```diff
@@ -128,12 +128,8 @@ func newNodeStore(nodeName string, conf Configuration, owner Owner, k8sEventReg
...
-						store.updateLocalNodeResource(newNode.DeepCopy())
+						equal = !store.updateLocalNodeResource(newNode.DeepCopy())
```
Comment (twpayne) [inline pkg/ipam/crd.go]: Overall, I found this change quite hard to understand, but I think a bit of renaming would help. Specifically:
* The name of the function`updateLocalNodeResource` does not indicate that it returns a value, yet here we a receiving a `bool` and inverting it.
*`updateLocalNodeResource` returns a `bool` which it calls `changed`, but here we invert it with `!` and call it `equal`.
* At line 120 in this file, we also call `updateLocalNodeResource` but ignore its return value.

`updateLocalNodeResource` is really doing multiple things: storing the latest node value, determining whether the node's spec or status has changed, and updating ENI routes and allocation pool sizes. I think the code should be refactored to at least move the change detection out of `updateLocalNodeResource`.

Comment (gandro) [reply pkg/ipam/crd.go]: Good feedback, thanks! I looked at the code again and found that introducing a side-effect free setter instead makes the change much easier to understand.

Please re-review.

**Review by twpayne — APPROVED** [review]
This is great, thanks!

**Comment by gandro** [comment]
ConformanceEKS (the only CI where the code modified here is enabled) failed with the connectivity check with IPSec, which is a known flake: https://github.com/cilium/cilium/issues/16938 https://github.com/cilium/cilium/actions/runs/1475670758

**Comment by gandro** [comment]
- https://jenkins.cilium.io/job/Cilium-PR-K8s-1.20-kernel-5.4/195/ hit #16852
- https://jenkins.cilium.io/job/Cilium-PR-K8s-1.21-kernel-4.19/196 and https://jenkins.cilium.io/job/Cilium-PR-K8s-1.22-kernel-4.9/186/hit a variants of #13839 (which is currently known to affect many tests in that test suite)

The modified code is only active in ENI mode (i.e. ConformanceEKS, which passed). The failures are therefore unrelated. Marking as ready-to-merge.

**Comment by qmonnet** [comment]
@gandro Can you also have a look at the failure on `k8s-1.22-kernel-4.9` please?

**Comment by gandro** [comment]
@qmonnet Apologies, on first glance it looked like a variant of #13839 as mentioned above, but might be a new flake since the symptoms are different (the affected tests are the same).

Will take a closer look and open an issue and then mark this again, but I'm very confident that this is unrelated to the PR since the code modified by this PR is not active in our Jenkins suite at all.

**Comment by gandro** [comment]
The link for https://jenkins.cilium.io/job/Cilium-PR-K8s-1.22-kernel-4.9/186/ just expired (it was still valid a few hours ago) :grimacing: Making it impossible to triage it again. Will restart it.

**Comment by gandro** [comment]
/test-1.22-4.9

**Comment by qmonnet** [comment]
Based on Sebastian's previous analysis & confidence and on the fact that the new run passed, I'll go ahead and merge. Thanks!

**Comment by gandro** [comment]
Marking this for backport to v1.10. Seems like users are hitting this on v1.10, and the backport should be rather trivial.

**Pull request merged**